File: sierpinski_curve/sierpinski_curve.cpp
// See https://en.wikipedia.org/wiki/Sierpi%C5%84ski_curve#Representation_as_Lindenmayer_system
#include <cmath>
#include <fstream>
#include <iostream>
#include <string>

std::string rewrite(const std::string& s) {
    std::string t;
    for (char c : s) {
        if (c == 'X')
            t += "XF+G+XF--F--XF+G+X";
        else
            t += c;
    }
    return t;
}

void line(std::ostream& out, double& x, double& y, double length, int angle) {
    constexpr double pi = 3.14159265359;
    double theta = (pi * angle)/180.0;
    x += length * std::cos(theta);
    y += length * std::sin(theta);
    out << 'L' << x << ',' << y << '\n';
}

void execute(std::ostream& out, const std::string& s, double x, double y,
             double length, int angle) {
    out << 'M' << x << ',' << y << '\n';
    for (char c : s) {
        if (c == 'F' || c == 'G')
            line(out, x, y, length, angle);
        else if (c == '+')
            angle = (angle - 45) % 360;
        else if (c == '-')
            angle = (angle + 45) % 360;
    }
}

int main() {
    const int size = 545;
    const int order = 5;
    const double x = 535, y = 5, length = 7;
    std::ofstream out("sierpinski_curve.svg");
    if (!out) {
        std::cerr << "Cannot open output file\n";
        return 1;
    }
    out << "<svg xmlns='http://www.w3.org/2000/svg' width='"
        << size << "' height='" << size << "'>\n";
    out << "<rect width='100%' height='100%' fill='white'/>\n";
    out << "<path stroke-width='1' stroke='black' fill='none' d='";
    std::string s = "F--XF--F--XF";
    for (int i = 0; i < order; ++i)
        s = rewrite(s);
    execute(out, s, x, y, length, 45);
    out << "'/>\n</svg>\n";
    return 0;
}

File: sierpinski_curve/sierpinski_curve.rs
// [dependencies]
// svg = "0.8.0"

use svg::node::element::path::Data;
use svg::node::element::Path;

struct SierpinskiCurve {
    current_x: f64,
    current_y: f64,
    current_angle: i32,
    line_length: f64,
}

impl SierpinskiCurve {
    fn new(x: f64, y: f64, length: f64, angle: i32) -> SierpinskiCurve {
        SierpinskiCurve {
            current_x: x,
            current_y: y,
            current_angle: angle,
            line_length: length,
        }
    }
    fn rewrite(order: usize) -> String {
        let mut str = String::from("F--XF--F--XF");
        for _ in 0..order {
            let mut tmp = String::new();
            for ch in str.chars() {
                match ch {
                    'X' => tmp.push_str("XF+G+XF--F--XF+G+X"),
                    _ => tmp.push(ch),
                }
            }
            str = tmp;
        }
        str
    }
    fn execute(&mut self, order: usize) -> Path {
        let mut data = Data::new().move_to((self.current_x, self.current_y));
        for ch in SierpinskiCurve::rewrite(order).chars() {
            match ch {
                'F' => data = self.draw_line(data),
                'G' => data = self.draw_line(data),
                '+' => self.turn(45),
                '-' => self.turn(-45),
                _ => {}
            }
        }
        Path::new()
            .set("fill", "none")
            .set("stroke", "black")
            .set("stroke-width", "1")
            .set("d", data)
    }
    fn draw_line(&mut self, data: Data) -> Data {
        let theta = (self.current_angle as f64).to_radians();
        self.current_x += self.line_length * theta.cos();
        self.current_y -= self.line_length * theta.sin();
        data.line_to((self.current_x, self.current_y))
    }
    fn turn(&mut self, angle: i32) {
        self.current_angle = (self.current_angle + angle) % 360;
    }
    fn save(file: &str, size: usize, order: usize) -> std::io::Result<()> {
        use svg::node::element::Rectangle;
        let x = 5.0;
        let y = 10.0;
        let rect = Rectangle::new()
            .set("width", "100%")
            .set("height", "100%")
            .set("fill", "white");
        let mut s = SierpinskiCurve::new(x, y, 7.0, 45);
        let document = svg::Document::new()
            .set("width", size)
            .set("height", size)
            .add(rect)
            .add(s.execute(order));
        svg::save(file, &document)
    }
}

fn main() {
    SierpinskiCurve::save("sierpinski_curve.svg", 545, 5).unwrap();
}
